File: document/src/vespa/document/fieldvalue/boolfieldvalue.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "boolfieldvalue.h"
#include <vespa/document/datatype/datatype.h>
#include <vespa/vespalib/util/xmlstream.h>
#include <ostream>

using namespace vespalib::xml;

namespace document {

BoolFieldValue::BoolFieldValue(bool value)
    : FieldValue(Type::BOOL), _value(value) {
}

BoolFieldValue::~BoolFieldValue() = default;

FieldValue &
BoolFieldValue::assign(const FieldValue &rhs) {
    if (rhs.isA(Type::BOOL)) {
        operator=(static_cast<const BoolFieldValue &>(rhs));
        return *this;
    } else {
        return FieldValue::assign(rhs);
    }
}

int
BoolFieldValue::compare(const FieldValue&rhs) const {
    int diff = FieldValue::compare(rhs);
    if (diff != 0) return diff;
    const BoolFieldValue &o = static_cast<const BoolFieldValue &>(rhs);
    return (_value == o._value) ? 0 : _value ? 1 : -1;
}

void
BoolFieldValue::printXml(XmlOutputStream& out) const {
    out << XmlContent(getAsString());
}

void
BoolFieldValue::print(std::ostream& out, bool, const std::string&) const {
    out << (_value ? "true" : "false") << "\n";
}

const DataType *
BoolFieldValue::getDataType() const {
    return DataType::BOOL;
}

FieldValue *
BoolFieldValue::clone() const {
    return new BoolFieldValue(*this);
}

char
BoolFieldValue::getAsByte() const {
    return _value ? 1 : 0;
}
int32_t
BoolFieldValue::getAsInt() const {
    return _value ? 1 : 0;
}
int64_t
BoolFieldValue::getAsLong() const {
    return _value ? 1 : 0;
}
float
BoolFieldValue::getAsFloat() const {
    return _value ? 1 : 0;
}
double
BoolFieldValue::getAsDouble() const {
    return _value ? 1 : 0;
}
vespalib::string
BoolFieldValue::getAsString() const {
    return _value ? "true" : "false";
}

BoolFieldValue&
BoolFieldValue::operator=(vespalib::stringref v) {
    _value = (v == "true");
    return *this;
}

}  // namespace document

File: document/src/main/java/com/yahoo/document/datatypes/BoolFieldValue.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document.datatypes;

import com.yahoo.document.DataType;
import com.yahoo.document.Field;
import com.yahoo.document.PrimitiveDataType;
import com.yahoo.document.serialization.FieldReader;
import com.yahoo.document.serialization.FieldWriter;
import com.yahoo.document.serialization.XmlSerializationHelper;
import com.yahoo.document.serialization.XmlStream;
import com.yahoo.vespa.objects.Ids;

/**
 * A boolean field value
 *
 * @author bratseth
 */
public class BoolFieldValue extends FieldValue {

    private static class Factory extends PrimitiveDataType.Factory {
        @Override public FieldValue create() { return new BoolFieldValue(); }
        @Override public FieldValue create(String value) { return new BoolFieldValue(value); }
    }

    public static PrimitiveDataType.Factory getFactory() { return new Factory(); }
    public static final int classId = registerClass(Ids.document + 17, BoolFieldValue.class);
    private boolean value;

    public BoolFieldValue() {
        this(false);
    }

    public BoolFieldValue(boolean value) {
        this.value = value;
    }

    public BoolFieldValue(String s) { value = Boolean.parseBoolean(s); }

    @Override
    public BoolFieldValue clone() {
        return (BoolFieldValue)super.clone();
    }

    @Override
    public void clear() {
        value = false;
    }

    @Override
    public void assign(Object o) {
        if ( ! checkAssign(o)) return;
        if (o instanceof String || o instanceof StringFieldValue) {
            value = Boolean.parseBoolean(o.toString());
        } else if (o instanceof Boolean) {
            value = (Boolean) o;
        } else if (o instanceof BoolFieldValue) {
            value = ((BoolFieldValue) o).value;
        } else {
            throw new IllegalArgumentException("Class " + o.getClass() + " not applicable to an " + this.getClass() + " instance.");
        }
    }

    public boolean getBoolean() {
        return value;
    }
    public void setBoolean(boolean value) { this.value = value; }

    @Override
    public Object getWrappedValue() {
        return value;
    }

    @Override
    public DataType getDataType() {
        return DataType.BOOL;
    }

    @Override
    public void printXml(XmlStream xml) {
        XmlSerializationHelper.printBoolXml(this, xml);
    }

    @Override
    public String toString() {
        return "" + value;
    }

    @Override
    public int hashCode() {
        return super.hashCode() + ( value ? 3 : 0);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if ( ! (o instanceof BoolFieldValue)) return false;
        if ( ! super.equals(o)) return false;

        BoolFieldValue that = (BoolFieldValue) o;
        return (value == that.value);
    }

    @Override
    public void serialize(Field field, FieldWriter writer) {
        writer.write(field, this);
    }

    /* (non-Javadoc)
     * @see com.yahoo.document.datatypes.FieldValue#deserialize(com.yahoo.document.Field, com.yahoo.document.serialization.FieldReader)
     */
    @Override
    public void deserialize(Field field, FieldReader reader) {
        reader.read(field, this);
    }

    @Override
    public int compareTo(FieldValue other) {
        int comp = super.compareTo(other);
        if (comp != 0) return comp;
        return Boolean.compare(value, ((BoolFieldValue)other).value);
    }

}
